### PR TITLE
ID-891 Optimize UserResourcesQuery

### PR DESCRIPTION
**Pull request by Ghost-in-a-Jar**
Ticket: https://broadworkbench.atlassian.net/browse/ID-891

What/why:

This query is accounting for the majority of the cpu load in sam, for some reason postgres is doing a full table scan of `sam_effective_resource_policy` every single time the query is ran, despite there being indexes on the table. This modification forces the join of `sam_effective_resource_policy` to filter down to only resource_ids of the type we are looking for.

How:

This modification to the query should be functionally equivalent, but forces the postgres query planner to do an index scan instead of a full table scan:

before
->  Parallel Seq Scan on sam_effective_resource_policy effpol  (cost=0.00..14559.75 rows=537375 width=24)
after
->  Index Scan using idx_erp_resource_id on sam_effective_resource_policy effpol  (cost=0.43..0.63 rows=6 width=24)

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala [hunk 1]
```diff
@@ -1781,6 +1781,11 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
           from ${ResourceTable as resource}
           join ${ResourceTypeTable as resourceType} on ${resource.resourceTypeId} = ${resourceType.id}
           join ${EffectiveResourcePolicyTable as effPol} on ${resource.id} = ${effPol.resourceId}
+            and ${effPol.resourceId} in (
+              select ${resource.id} from ${ResourceTable as resource}
+              join ${ResourceTypeTable as resourceType} on ${resourceType.id} = ${resource.resourceTypeId}
+              where ${resourceType.name} = ${resourceTypeName}
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:1790]: if you want to go a little further
```suggestion
              where ${resource.resourceTypeId} = ${resourceTypePKsByName(resourceTypeName)}
```
also could remove the ResourceTypeTable join and replace
```
where ${resourceType.name} = ${resourceTypeName}
```
with
```
where ${resource.resourceTypeId} = ${resourceTypePKsByName(resourceTypeName)}
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala [hunk 2]
```diff
@@ -1781,6 +1781,11 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
           from ${ResourceTable as resource}
           join ${ResourceTypeTable as resourceType} on ${resource.resourceTypeId} = ${resourceType.id}
           join ${EffectiveResourcePolicyTable as effPol} on ${resource.id} = ${effPol.resourceId}
+            and ${effPol.resourceId} in (
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:1787]: drop a comment about why this is here

**Review by dvoet — APPROVED** [review]
made a suggestion for now or for the future

**Review by tlangs — COMMENTED** [review]
Lets test on Dev and make sure that the query is returning everything and behaving as we expect. Once that's done, to prod we go!

**Comment by Ghost-in-a-Jar** [comment]
@tlangs I ran:
```
(select * from optimized EXCEPT select * from original)
UNION ALL
(select * from original EXCEPT select * from optimized)
```
and that returned nothing so they are returning the same rows

**Review by tlangs — APPROVED** [review]

**Pull request merged**